File: src/main/scala/eu/unitn/disi/distance/metrics/AuctionAlg.scala
package eu.unitn.disi.distance.metrics

import eu.unitn.disi.distance.utilities.DistanceWrapper
import org.apache.spark.sql.functions.col
import org.apache.spark.sql.{Dataset, SparkSession}

import scala.collection.mutable

class AuctionAlg(spark: SparkSession,
                 distances: Dataset[DistanceWrapper],
                 dim: Int)
    extends IAlg(spark, distances, dim) {

  val p: Array[Double] = Array.fill(dim)(0)
  val owner: Array[Int] = Array.fill(dim)(-1)
  val epsilon: Double = 1.0 / (dim + 1)
  val queue: mutable.Queue[Int] = mutable.Queue[Int]()
  var distancesMap: Map[(Int, Int), DistanceWrapper] = _

  override def computeMatching(): Array[DistanceWrapper] = {
    val start = System.currentTimeMillis()
    var assignedA: Set[Int] = Set.empty[Int]
    var assignedB: Set[Int] = Set.empty[Int]

    val zeros = distances.filter(col("distance") === 0.0).collect()
    for (elem <- zeros) {
      assignedA = assignedA ++ Seq(elem.idA)
      assignedB = assignedB ++ Seq(elem.idB)
      matches = matches ++ Seq(elem)
    }

    val distancesList: Array[DistanceWrapper] = distances
      .filter(col("distance") =!= 0.0)
      .filter(
        !col("idA")
          .isin(zeros.map(_.idA): _*)
          .or(!col("idB").isin(zeros.map(_.idB): _*)))
      .collect()

    timeToCollect = System.currentTimeMillis() - start

    log.info(s"Collected distances in ${timeToCollect}ms")

    distancesMap = fromListToMap(distancesList)

    List
      .range(0, dim)
      .foreach(elem =>
        if (!assignedA.contains(elem)) {
          queue.enqueue(elem)
      })

    while (queue.nonEmpty) {
      val i = queue.dequeue()
      var max = Double.MinValue
      var maxIndex = -1
      for (j <- 0 until dim if !assignedB.contains(j)) {
        val dist =
          distancesMap.getOrElse((i, j), DistanceWrapper(i, j, 1.0)).distance
        if (max < (1.0 - dist - p(j))) {
          max = 1.0 - dist - p(j)
          maxIndex = j
        }
      }
      if (max >= 0) {
        if (owner(maxIndex) != -1) {
          queue.enqueue(owner(maxIndex))
        }
        owner(maxIndex) = i
        p(maxIndex) += epsilon
      }
    }

    for (idx <- owner.indices if !assignedB.contains(idx)) {
      if (!assignedA.contains(owner(idx))) {
        matches = matches ++ Seq(
          distancesMap.getOrElse((owner(idx), idx),
                                 DistanceWrapper(owner(idx), idx, 1.0)))
      }
    }
    matches
  }
}

File: dm/algorithms/AuctionAlg.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import numpy as np
from collections import deque
from pyspark.sql import functions as F
from .AbstractDistanceAlg import AbstractDistanceAlg


class AuctionAlg(AbstractDistanceAlg):

    def __init__(self, df, size):
        super().__init__(df, size)

    def compute_matching(self):
        fst_assigned = set()
        snd_assigned = set()
        p = np.full(self.size, 0)
        owner = np.full(self.size, -1)
        epsilon = 1.0 / (self.size + 1)

        zeros = self.df.filter(F.col('distance') == 0.0).collect()
        for row, col, dist in zeros:
            fst_assigned.add(row)
            snd_assigned.add(col)
            self.matches.append((row, col, dist))

        distances = self.df \
            .filter(F.col('distance') != 0.0) \
            .filter(not F.col('idA').isin([elem[0] for elem in zeros])) \
            .filter(not F.col('idB').isin([elem[0] for elem in zeros])) \
            .collect()

        distances_map = AuctionAlg.from_list_to_map(distances)
        queue = deque()
        for idx in range(0, self.size):
            if idx not in fst_assigned:
                queue.append(idx)

        while queue:
            row = queue.popleft()
            col_idx = -1
            col_value = float('inf')
            for col in range(0, self.size):
                if col not in snd_assigned:
                    distance = distances_map.get((row, col), (row, col, 1.0))[2]
                    curr_value = 1.0 - distance - p[0]
                    if col_value < curr_value:
                        col_value = curr_value
                        col_idx = col

            if col_value >= 0:
                if not owner[col_idx] == -1:
                    queue.append(owner[col_idx])
                owner[col_idx] = row
                p[col_idx] = p[col_idx] + epsilon

        for idx, elem in enumerate(owner):
            if idx not in snd_assigned:
                self.matches.append(distances_map.get((elem, idx), (elem, idx, 1.0)))

        return self.matches
